syntax = "proto2";

package blockchain;

// Dispersy messages
message SignatureRequestMessage {
    required uint32 identifier = 1;
    required Contract contract = 2;
}

message SignatureResponseMessage {
    required uint32 identifier = 1;
    required Contract contract = 2;
}

message ContractMessage {
    required Contract contract = 1;
}

message BlockRequestMessage {
    required bytes block_id = 1;
}

message BlockMessage {
    required Block block = 1;
}

message TraversalRequestMessage {
    required uint32 identifier = 1;
    required bytes contract_id = 2;
}

message TraversalResponseMessage {
    required uint32 identifier = 1;
    optional Contract contract = 2;
}

// Objects that are included in the Dispersy messages
message Contract {
    required bytes previous_hash = 1;
    required bytes from_public_key = 2;
    required bytes from_signature = 3;
    required bytes to_public_key = 4;
    required bytes to_signature = 5;
    required bytes document = 6;
    required uint32 type = 7;
    required uint32 time = 8;
}

message Block {
    required bytes previous_hash = 1;
    required bytes merkle_root_hash = 2;
    required bytes target_difficulty = 3;
    required bytes creator = 4;
    required bytes creator_signature = 5;
    required uint32 time = 6;
    repeated Contract contracts = 7;
}
